heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java

```diff
@@ -0,0 +1,404 @@
+        instancePlanMap.put(instanceId, instancePlan);
+      }
+
+      containerCpu += (paddingPercentage * containerCpu) / 100;
+      containerRam += (paddingPercentage * containerRam) / 100;
+      containerDiskInBytes += (paddingPercentage * containerDiskInBytes) / 100;
+
+      Resource resource =
+          new Resource(Math.round(containerCpu), containerRam, containerDiskInBytes);
+
+      PackingPlan.ContainerPlan containerPlan =
+          new PackingPlan.ContainerPlan(containerId, instancePlanMap, resource);
+
+      containerPlanMap.put(containerId, containerPlan);
+      topologyRam += containerRam;
+      topologyCpu += Math.round(containerCpu);
+      topologyDisk += containerDiskInBytes;
+    }
+
+    // Take the heron internal container into account and the application master for YARN
+    // scheduler
+    topologyRam += instanceRamDefault;
+    topologyDisk += instanceDiskDefault;
+    topologyCpu += instanceCpuDefault;
+
+    Resource resource = new Resource(
+        topologyCpu, topologyRam, topologyDisk);
+
+    PackingPlan plan = new PackingPlan(topology.getId(), containerPlanMap, resource);
+    return plan;
+  }
+
+  @Override
+  public void close() {
+  }
+
+  /**
+   * Get the RAM requirements of all the components
+   *
+   * @return The list of components and their RAM requirements. Returns null if one or more
+   * instances do not have valid resource requirements.
+   */
+  protected ArrayList<RamRequirement> getRAMInstances() {
+    ArrayList<RamRequirement> ramRequirements = new ArrayList<>();
+    Map<String, Integer> parallelismMap = TopologyUtils.getComponentParallelism(topology);
+    Map<String, Long> ramMap = TopologyUtils.getComponentRamMapConfig(topology);
+
+    for (String component : parallelismMap.keySet()) {
+      if (ramMap.containsKey(component)) {
+        if (!isValidInstance(new Resource(instanceCpuDefault,
+            ramMap.get(component), instanceDiskDefault))) {
+          return null;
+        } else {
+          ramRequirements.add(new RamRequirement(component, ramMap.get(component)));
+        }
+      } else {
+        if (!isValidInstance(new Resource(instanceCpuDefault,
+            instanceRamDefault, instanceDiskDefault))) {
+          return null;
+        } else {
+          ramRequirements.add(new RamRequirement(component, instanceRamDefault));
+        }
+      }
+    }
+    return ramRequirements;
+  }
+
+  /**
+   * Get the instances' allocation based on the Resource Compliant Round Robin algorithm
+   *
+   * @return Map &lt; containerId, list of InstanceId belonging to this container &gt;
+   */
+  protected Map<String, List<String>> getResourceCompliantRRAllocation() {
+
+    Map<String, List<String>> allocation = new HashMap<>();
+    ArrayList<Container> containers = new ArrayList<>();
+    ArrayList<RamRequirement> ramRequirements = getRAMInstances();
+
+    if (ramRequirements == null) {
+      throw new RuntimeException("The topology configuration does not have "
```

Comment:
If this exception were thrown the reader would have little info about what went wrong. Instead could getRAMInstances throw a more detailed exception instead of returning null to this method, which then throws an exception? Best to throw as soon as an erroneous condition is found, which in this case is within getRAMInstances.